### ISSUE TITLE
Competing catch-all routes break client-side navigation in dev mode

### ISSUE THREAD

**Issue by Parnswir** [vercel/next.js#54376]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: linux
  Arch: x64
  Version: #1 SMP PREEMPT_DYNAMIC Sun Aug  6 20:05:33 UTC 2023
Binaries:
  Node: 16.17.0
  npm: 8.15.0
  Yarn: 1.22.19
  pnpm: 7.1.0
Relevant Packages:
  next: 13.4.20-canary.1
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.1.3
Next.js Config:
  output: N/A
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

Routing (next/router, next/navigation, next/link)

### Link to the code that reproduces this issue or a replay of the bug

https://codesandbox.io/p/sandbox/friendly-antonelli-xpmzl4?file=/pages/index.tsx

### To Reproduce

1. Open [CodeSandbox](https://codesandbox.io/p/sandbox/friendly-antonelli-xpmzl4?file=/pages/index.tsx:12,1)
2. In dev mode: click on links on index page
  a. HTML link (<a href=>) shows intended behavior
  b. next/link component shows bug behavior

Please note, that the bug behavior only appears in dev mode (`npm run dev`).

### Describe the Bug

We have two routes, `/somepath/[id]/` and `/[...slug]`. The former is used to access server-side rendered pages for a collection of resources, the latter is a catch-all route with getStaticPaths (to render some pages from a CMS) and some custom 404 logic.

Since updating to `next@13.4.13`, we're now seeing an erroneous behavior in dev mode when using client-side navigation to paths like `/somepath/foo`. Instead of the correct page, we see the 404 logic of the `/[...slug]`-catch-all page being executed and leading to rendering the 404. This behavior is not exhibited when running `next build && next start`.

The issue occurrs only if `/somepath/[id]/` uses `getServerSideProps` and if `/[...slug]`'s `getStaticProps` returns `{ notFound: true }`. This could mean there is some kind of mismatch in how `getStaticPaths` is handled in the dev- and production-server respectively. Maybe this is related to https://github.com/vercel/next.js/commit/1398de9977b89c9c4717d26e213b52bc63ccbe7e (which is part of the breaking patch release 13.4.13)?

### Expected Behavior

Navigation should work as expected in dev mode.

**Comment by samipshah100** [comment]
I am having the same issue.

**Comment by marcinincreo** [comment]
Same issue for me. This prevents us from further updating all NextJS solutions.

**Comment by marcinincreo** [comment]
@Parnswir I've checked the 13.4.20-canary.9 version and it seems that the issue is not there anymore.

**Comment by timneutkens** [comment]
This was fixed in #54364
